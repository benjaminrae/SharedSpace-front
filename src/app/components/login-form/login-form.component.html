<form [formGroup]="loginForm" (ngSubmit)="onSubmit()">
  <h2>Log in to SharedSpace</h2>

  <div class="form-group">
    <label for="username">Username</label>
    <input id="username" type="text" formControlName="username">
  </div>

  <div class="form-group">
    <label for="password">Password</label>
    <input id="password" type="password" formControlName="password">
  </div>

  <app-button text="Login" [big]="true" [disabled]="loginForm.status === 'INVALID'"></app-button>

  <span>Don't have an account? <a href="">Register here</a> </span>

</form>
