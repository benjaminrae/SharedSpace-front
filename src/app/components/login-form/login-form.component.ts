import { Component } from "@angular/core";
import { Validators, FormBuilder } from "@angular/forms";
import { TokenService } from "src/app/services/token/token.service";
import { UiService } from "src/app/services/ui/ui.service";
import { UserService } from "src/app/services/user/user.service";
import { UserCredentials } from "src/app/store/user-feature/types";

@Component({
  selector: "app-login-form",
  templateUrl: "./login-form.component.html",
})
export class LoginFormComponent {
  loginForm = this.fb.group({
    username: ["", [Validators.required, Validators.minLength(5)]],
    password: ["", [Validators.required, Validators.minLength(8)]],
  });

  constructor(
    private readonly fb: FormBuilder,
    private readonly userService: UserService,
    private readonly uiService: UiService,
    private readonly tokenService: TokenService
  ) {}

  onSubmit() {
    this.uiService.showLoading();

    const data = this.userService.getToken(
      this.loginForm.value as UserCredentials
    );

    data.subscribe((data) => {
      const { token } = data;
      const { id, username } = this.tokenService.decodeToken(token);
      this.userService.loginUser({ username, id, token });
      this.uiService.hideLoading();
      this.resetForm();
      this.uiService.showSuccessModal("You have logged in correctly");
    });
  }

  private resetForm() {
    this.loginForm.reset();
  }
}
