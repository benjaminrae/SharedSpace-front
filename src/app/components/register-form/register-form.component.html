<form [formGroup]="registerForm" (ngSubmit)="onSubmit()" data-testid="form">
  <h2>Sign up for SharedSpace</h2>

  <div class="form-group">
    <label for="username">Username</label>
    <input id="username" type="text" formControlName="username" required autocomplete="off">
  </div>

  <div class="form-group">
    <label for="password">Password</label>
    <input id="password" type="password" formControlName="password" required autocomplete="off">
  </div>

  <div class="form-group">
    <label for="confirmPassword">Confirm password</label>
    <input id="confirmPassword" type="password" formControlName="confirmPassword" required autocomplete="off">
  </div>

  <div class="form-group ">
    <label for="owner"><input type="checkbox" name="owner" id="owner" formControlName="owner">I want to offer a
      space</label>
  </div>

  <app-button text="Sign up" [big]="true" [disabled]="registerForm.status === 'INVALID'"></app-button>

  <span>Already have an account?
    <a routerLink="/login">Login here</a>
  </span>

</form>
