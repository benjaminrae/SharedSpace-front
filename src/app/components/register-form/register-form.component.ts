import { Component } from "@angular/core";
import { FormBuilder, Validators } from "@angular/forms";
import { UiService } from "../../services/ui/ui.service";
import { RegisterUserCredentials } from "../../store/user-feature/types";
import { UserService } from "../../services/user/user.service";

@Component({
  selector: "app-register-form",
  templateUrl: "./register-form.component.html",
})
export class RegisterFormComponent {
  registerForm = this.fb.group({
    username: ["", [Validators.required, Validators.minLength(5)]],
    password: ["", [Validators.required, Validators.minLength(8)]],
    confirmPassword: ["", [Validators.required, Validators.minLength(8)]],
    owner: [false],
  });

  constructor(
    private readonly fb: FormBuilder,
    private readonly userService: UserService,
    private readonly uiService: UiService
  ) {}

  onSubmit() {
    this.uiService.showLoading();

    const data = this.userService.registerUser(
      this.registerForm.value as RegisterUserCredentials
    );

    data.subscribe((data) => {
      this.uiService.hideLoading();
      this.uiService.showSuccessModal("You have successfully registered");
      this.resetForm();
    });
  }

  private resetForm() {
    this.registerForm.reset();
  }
}
