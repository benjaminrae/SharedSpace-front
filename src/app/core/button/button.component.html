<button [routerLink]="routerLink" [type]="type" [className]="className" (click)="onClickButton($event)"
  [disabled]="disabled" #button>{{text}}</button>
