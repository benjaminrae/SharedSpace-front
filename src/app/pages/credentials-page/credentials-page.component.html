<div class="credentials-page">
  <img src="../../../assets/svgs/workspace.svg" alt="A man working at a desk" #image />

  <app-login-form class="credentials-page__form" *ngIf="page === '/login'"></app-login-form>
  <app-register-form class="credentials-page__form" *ngIf="page === '/register'"></app-register-form>
</div>
