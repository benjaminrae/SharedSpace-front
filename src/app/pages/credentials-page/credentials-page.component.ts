import { Component, Input, OnInit } from "@angular/core";
import { Router } from "@angular/router";

@Component({
  selector: "app-credentials-page",
  templateUrl: "./credentials-page.component.html",
  styleUrls: ["./credentials-page.component.scss"],
})
export class CredentialsPageComponent implements OnInit {
  page!: "/login" | "/register";

  constructor(private readonly router: Router) {}

  ngOnInit(): void {
    this.page = this.router.url as "/login" | "/register";
  }
}
