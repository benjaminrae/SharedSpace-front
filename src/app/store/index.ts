import { type ActionReducerMap, type MetaReducer } from "@ngrx/store";
import { environment } from "../../environments/environment";
import { locationsFeature } from "./locations-feature/locations-feature.reducer";
import { type ApplicationState } from "./types";
import { uiFeature } from "./ui-feature/ui-feature.reducer";
import { userFeature } from "./user-feature/user-feature.reducer";

export const reducers: ActionReducerMap<ApplicationState> = {
  ui: uiFeature.reducer,
  user: userFeature.reducer,
  locations: locationsFeature.reducer,
};

export const metaReducers: Array<MetaReducer<ApplicationState>> =
  !environment.production ? [] : [];
