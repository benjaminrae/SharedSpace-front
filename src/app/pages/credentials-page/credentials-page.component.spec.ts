import { HttpClient, HttpHandler } from "@angular/common/http";
import { ComponentFixture, TestBed } from "@angular/core/testing";
import { FormBuilder } from "@angular/forms";
import { provideMockStore } from "@ngrx/store/testing";
import { render, screen } from "@testing-library/angular";
import { ButtonComponent } from "../../core/button/button.component";
import { LoginFormComponent } from "../../components/login-form/login-form.component";
import { RouterTestingModule } from "@angular/router/testing";
import { CredentialsPageComponent } from "./credentials-page.component";
import { routes } from "../../app-routing.module";
import { Router } from "@angular/router";
import { RegisterFormComponent } from "../../components/register-form/register-form.component";

const providers = [FormBuilder, HttpClient, HttpHandler, provideMockStore({})];
const declarations = [
  RegisterFormComponent,
  LoginFormComponent,
  ButtonComponent,
];

describe("Given the CredentialsPageComponent", () => {
  let component: CredentialsPageComponent;
  let fixture: ComponentFixture<CredentialsPageComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [CredentialsPageComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(CredentialsPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  describe("When it is rendered", () => {
    test("Then the component should be created", () => {
      expect(component).toBeTruthy();
    });
  });
});

describe("Given the CredentialsPage component", () => {
  describe("When it is rendered and the path is /login", () => {
    test("Then it should match the snapshot", async () => {
      const { fixture, detectChanges } = await render(
        CredentialsPageComponent,
        {
          declarations,
          providers,
          imports: [RouterTestingModule.withRoutes(routes)],
          detectChanges: false,
        }
      );

      const router = TestBed.inject(Router);

      await fixture.ngZone!.run(async () => router.navigate(["/login"]));

      detectChanges();

      expect(fixture).toMatchSnapshot();
    });

    test("And then it should show an image on the screen with name 'A man working at a desk' and a heading level 2 with 'Log in to SharedSpace'", async () => {
      const imageAltText = "A man working at a desk";
      const loginHeading = {
        level: 2,
        name: "Log in to SharedSpace",
      };

      const { fixture, detectChanges } = await render(
        CredentialsPageComponent,
        {
          declarations,
          providers,
          imports: [RouterTestingModule.withRoutes(routes)],
          detectChanges: false,
        }
      );

      const router = TestBed.inject(Router);

      await fixture.ngZone!.run(async () => router.navigate(["/login"]));

      detectChanges();

      const renderedHeading = screen.queryByRole("heading", loginHeading);
      const renderedImage = screen.queryByRole("img", {
        name: imageAltText,
      });

      expect(renderedHeading).toBeInTheDocument();
      expect(renderedImage).toBeInTheDocument();
    });
  });

  describe("When it is rendered and the path is /register", () => {
    test("Then it should match the snapshot", async () => {
      const { fixture, detectChanges } = await render(
        CredentialsPageComponent,
        {
          declarations,
          providers,
          imports: [RouterTestingModule.withRoutes(routes)],
          detectChanges: false,
        }
      );

      const router = TestBed.inject(Router);

      await fixture.ngZone!.run(async () => router.navigate(["/register"]));

      detectChanges();

      expect(fixture).toMatchSnapshot();
    });

    test("And then it should show an image on the screen with name 'A man working at a desk' and a heading level 2 with 'Sign up for SharedSpace'", async () => {
      const imageAltText = "A man working at a desk";
      const signupHeading = {
        level: 2,
        name: "Sign up for SharedSpace",
      };

      const { fixture, detectChanges } = await render(
        CredentialsPageComponent,
        {
          declarations,
          providers,
          imports: [RouterTestingModule.withRoutes(routes)],
          detectChanges: false,
        }
      );

      const router = TestBed.inject(Router);

      await fixture.ngZone!.run(async () => router.navigate(["/register"]));

      detectChanges();

      const renderedHeading = screen.queryByRole("heading", signupHeading);
      const renderedImage = screen.queryByRole("img", {
        name: imageAltText,
      });

      expect(renderedHeading).toBeInTheDocument();
      expect(renderedImage).toBeInTheDocument();
    });
  });
});
