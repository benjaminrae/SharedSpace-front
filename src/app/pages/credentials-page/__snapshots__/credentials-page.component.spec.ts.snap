// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the CredentialsPage component When it is rendered and the path is /login Then it should match the snapshot 1`] = `
<app-credentials-page
  __ngContext__={[Function Number]}
  page={[Function String]}
  router={[Function Router]}
>
  <div
    class="credentials-page"
  >
    <img
      alt="A man working at a desk"
      src="../../../assets/svgs/workspace.svg"
    />
    <app-login-form
      class="credentials-page__form"
    >
      <form
        data-testid="form"
      >
        <h2>
          Log in to SharedSpace
        </h2>
        <div
          class="form-group"
        >
          <label
            for="username"
          >
            Username
          </label>
          <input
            autocomplete="off"
            formcontrolname="username"
            id="username"
            required=""
            type="text"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="password"
          >
            Password
          </label>
          <input
            autocomplete="off"
            formcontrolname="password"
            id="password"
            required=""
            type="password"
          />
        </div>
        <app-button
          text="Login"
        >
          <button
            class="button button--big"
            disabled=""
            type="submit"
          >
            Login
          </button>
        </app-button>
        <span>
          Don't have an account? 
          <a
            href="/register"
            routerlink="/register"
          >
            Register here
          </a>
        </span>
      </form>
    </app-login-form>
  </div>
</app-credentials-page>
`;

exports[`Given the CredentialsPage component When it is rendered and the path is /register Then it should match the snapshot 1`] = `
<app-credentials-page
  __ngContext__={[Function Number]}
  page={[Function String]}
  router={[Function Router]}
>
  <div
    class="credentials-page"
  >
    <img
      alt="A man working at a desk"
      src="../../../assets/svgs/workspace.svg"
    />
    <app-register-form
      class="credentials-page__form"
    >
      <form
        data-testid="form"
      >
        <h2>
          Sign up for SharedSpace
        </h2>
        <div
          class="form-group"
        >
          <label
            for="username"
          >
            Username
          </label>
          <input
            autocomplete="off"
            formcontrolname="username"
            id="username"
            required=""
            type="text"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="password"
          >
            Password
          </label>
          <input
            autocomplete="off"
            formcontrolname="password"
            id="password"
            required=""
            type="password"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="confirmPassword"
          >
            Confirm password
          </label>
          <input
            autocomplete="off"
            formcontrolname="confirmPassword"
            id="confirmPassword"
            required=""
            type="password"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="owner"
          >
            <input
              formcontrolname="owner"
              id="owner"
              name="owner"
              type="checkbox"
            />
            I want to offer a space
          </label>
        </div>
        <app-button
          text="Sign up"
        >
          <button
            class="button button--big"
            disabled=""
            type="submit"
          >
            Sign up
          </button>
        </app-button>
        <span>
          Already have an account? 
          <a
            href="/login"
            routerlink="/login"
          >
            Login here
          </a>
        </span>
      </form>
    </app-register-form>
  </div>
</app-credentials-page>
`;
